### DISCUSSION TITLE
主从架构下消息是如何保证不重复消费的

### DISCUSSION THREAD

**Discussion by jixiang-yz** [apache/rocketmq#6949] [Q&A]
假如一条消息在master 和slave 中都存在，那么如果有两个 同时 consumer 一个请求了 master 一个请求了slave ，这种情况下是如何保证消息不被重复消费的。

**Comment by mxsm** [comment]
消息的消费只会从Master进行消费， 但是当Master 宕机，在5.0版本下Controller模式的高可用模式会选举Slave作为master继续进行。而5.0之前的版本会从Slave继续进行消费， 并不会出现你上述说的一个请求了Master一个请求了Slave的情况。
5.0的主从高可用模式可用参考这个设计文档：https://github.com/apache/rocketmq/blob/develop/docs/cn/controller/design.md

**Reply by jixiang-yz** [reply]
那就是对于MQ来说没有所谓的读写分离，无论是消费还是生产实时上都是写操作，那也就是说slave相当于备机，而不是从机？

**Comment by mxsm** [comment]
> 那就是对于MQ来说没有所谓的读写分离，无论是消费还是生产实时上都是写操作，那也就是说slave相当于备机，而不是从机？

是的，4.x的版本你可以试Master-Slave模式将master干掉留下Slave，然后发送消息看一下。一般的情况下我们生产部署的方案都是多个Master-slave的模式不太可能出现所有的Master都宕机，整体来看集群是能够正常使用，但是Master宕机的只能消费消息，而不能接收消息。5.0 Controller模式会将Slave自动提升Master。所以能够继续使用。达到高可用。
